File: src/com/kraftwerking/datastructuresalgo/leetcode/WordBreak139.java
package com.kraftwerking.datastructuresalgo.leetcode;


import java.util.ArrayList;
import java.util.List;

class WordBreak139 {

  public boolean wordBreak(String s, List<String> wordDict) {
    int n = s.length();
    boolean[] dp = new boolean[n + 1];
    int maxLen = 0;
    for (String word : wordDict) {
      maxLen = Math.max(maxLen, word.length());
    }
    dp[0] = true;

    for (int i = 0; i <= n; i++) {
      for (int j = i - 1; j >= 0; j--) {
        if (i - j > maxLen) {
          continue;
        }
        if (dp[j] && wordDict.contains(s.substring(j, i))) {
          dp[i] = true;
          break;

        }
      }
    }

    return dp[n];
  }

  public static void main(String[] args) {
    WordBreak139 wordBreak139 = new WordBreak139();
    List<String> wordDict = new ArrayList<>();
    wordDict.add("leet");
    wordDict.add("code");
    boolean result = wordBreak139.wordBreak("leetcode", wordDict);
    System.out.println(result);
  }
}

File: python/leetcode/WordBreak139.py
from typing import List


class Solution(object):
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        dp = [False] * (len(s) + 1)
        dp[len(s)] = True
        for i in range(len(s) - 1, -1, -1):
            for w in wordDict:
                if (i + len(w)) <= len(s) and s[i: i + len(w)] == w:
                    dp[i] = dp[i + len(w)]
                    if dp[i]:
                        break

        return dp[0]


def main():
    s = "leetcode"
    wordDict = ["leet", "code"]
    solution = Solution()
    result = solution.wordBreak(s, wordDict)
    print(result)


if __name__ == "__main__":
    main()
